src/keyboardNavigation.ts

```diff
@@ -0,0 +1,83 @@
+import { isDeepestVisible, isImmediateParent } from './utils';
+
+function findDeepestVisibleChild(node: HTMLElement): HTMLElement {
+    if (isDeepestVisible(node)) return node;
+    const ul = <HTMLElement>node.children[1];
+    return findDeepestVisibleChild(<HTMLElement>ul.children[ul.children.length - 1]);
+}
+
+function findNearestParentWithAdditionalLeaves(node: HTMLElement): HTMLElement {
+    const ul = node.parentElement;
+    if (!ul) return node;
+    const parent = ul.parentElement;
+    if (!parent || !parent.parentElement) return node;
+    const parentAndSiblings = [...parent.parentElement.children];
+    if (parentAndSiblings.indexOf(parent) == parentAndSiblings.length - 1) {
+        return findNearestParentWithAdditionalLeaves(parent);
+    } else {
+        return <HTMLElement>parentAndSiblings[parentAndSiblings.indexOf(parent) + 1];
+    }
+}
+
+export const navigateUp = (path: string, shadowRoot: ShadowRoot): void => {
+    try {
+        const currentFocus = <HTMLElement>shadowRoot.querySelector(`[data-path="${path}"]`);
+        if (!currentFocus || !currentFocus.parentNode) return;
+        const currentFocusAndSiblings = [...currentFocus.parentNode.children];
+        var previousSibling: HTMLElement = <HTMLElement>(
+            currentFocus.parentNode.children[currentFocusAndSiblings.indexOf(currentFocus) - 1]
+        );
+        if (!previousSibling) {
+            //deal with the item being the first entry in a UL
+            previousSibling = <HTMLElement>currentFocus.parentNode.parentNode;
+        }
+        if (
+            !isDeepestVisible(previousSibling) &&
+            !isImmediateParent(path, previousSibling.getAttribute('data-path') || '')
+        ) {
+            changeFocusedElement(findDeepestVisibleChild(previousSibling), currentFocus);
+        } else {
+            changeFocusedElement(previousSibling, currentFocus);
+        }
+    } catch (err) {
+        console.error('Failed to navigate up via key press');
+    }
+};
+
+export const navigateDown = (path: string, shadowRoot: ShadowRoot): void => {
+    try {
+        const currentFocus = <HTMLElement>shadowRoot.querySelector(`[data-path="${path}"]`);
+        if (!currentFocus || !currentFocus.parentNode) return;
+        const currentFocusAndSiblings = [...currentFocus.parentNode.children];
+        if (!isDeepestVisible(currentFocus)) {
+            changeFocusedElement(<HTMLElement>currentFocus.children[1].children[0], currentFocus);
```

Comment:
i don't have great suggestions, but there are a few spots like this where changing the render logic a bit would break keyboard navigation, and i think it'd be really hard to repair because it requires thinking really hard to confirm that a given ancestry chain makes sense. just something to think about as we tailor

Reply:
I think I've made this somewhat more resilient by using querySelector instead of accessing arrays of children, but I still think if the rendering logic changes sufficiently it'll still break